File: java-playground/src/test/java/org/gonza/javaplayground/core/CarTest.java
package org.gonza.javaplayground.core;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.assertj.core.api.Assertions.assertThat;

public class CarTest {

    private Car car;

    @BeforeEach
    void setUp() {
        car = new Car("hong");
    }

    @Test
    @DisplayName("자동차가 생성되면 위치가 0이다.")
    void createCarSuccessTest() throws Exception {
        assertThat(car.getNickName()).isEqualTo("hong");
        assertThat(car.getPosition().getDistance()).isEqualTo(0);
    }

    @Test
    @DisplayName("랜덤 숫자가 4이상이면 한 칸 전진한다.")
    void forwardSuccessTest() throws Exception {
        int randomNumber = 4;

        car.forward(randomNumber);

        assertThat(car.getPosition().getDistance()).isEqualTo(1);
    }

    @ParameterizedTest
    @ValueSource(ints = {0, 1, 2, 3, -3, -2})
    @DisplayName("랜덤 숫자가 4미만이면 제자리이다.")
    void forwardFailTest_randomNumberIsUnder4(int invalidValue) throws Exception {

        car.forward(invalidValue);

        assertThat(car.getPosition().getDistance()).isEqualTo(0);
    }
}

File: kotlin-playground/src/test/kotlin/org/gonza/kotlinplayground/domain/car/CarTest.kt
package org.gonza.kotlinplayground.domain.car

import org.assertj.core.api.Assertions.*
import org.gonza.kotlinplayground.domain.car.exception.InvalidCarNameLengthException
import org.junit.jupiter.api.Test

class CarTest {
    @Test
    fun `자동차는 전진하면 위치가 변경된다`() {
        val initPosition = 0
        val initName = "name"
        val car = Car(initName, initPosition)

        val movedCar = car.move()

        assertThat(initPosition).isNotEqualTo(movedCar.position)
        assertThat(movedCar.position).isEqualTo(initPosition + 1)
    }

    @Test
    fun `자동차의 이름과 위치가 같은 두 자동차는 동일한 자동차이다`() {
        val car1 = Car("name", 1)
        val car2 = Car("name", 1)
        val otherCar = Car("name", 2)

        assertThat(car1).isEqualTo(car2)
        assertThat(car1).isNotEqualTo(otherCar)
        assertThat(car2).isNotEqualTo(otherCar)
    }

    @Test
    fun `자동차의 이름은 5글자를 초과할 수 없다`() {
        assertThatThrownBy {
            Car("123456", 1)
        }.isInstanceOf(InvalidCarNameLengthException::class.java)
    }
}
